sdk/keyvault/keyvault-certificates/src/core/challengeBasedAuthenticationPolicy.ts

```diff
@@ -82,102 +108,126 @@ export class ChallengeBasedAuthenticationPolicy extends BaseRequestPolicy {
-        }
-      }
+  /**
+   * Gets or updates the token from the token cache into the headers of the received web resource.
+   */
+  private async loadToken(webResource: WebResource): Promise<void> {
+    let accessToken = this.tokenCache.getCachedToken();
+
+    // If there's no cached token in the cache, we try to get a new one.
+    if (accessToken === undefined) {
+      const receivedToken = await this.credential.getToken(this.challengeCache.challenge!.scope);
+      accessToken = receivedToken || undefined;
+      this.tokenCache.setCachedToken(accessToken);
+    }
+
+    if (accessToken) {
+      webResource.headers.set(
+        Constants.HeaderConstants.AUTHORIZATION,
+        `Bearer ${accessToken.token}`
+      );
     }
-    return returnValue;
   }
 
   /**
-   * Applies the Bearer token to the request through the Authorization header.
-   * @param webResource
+   * Parses an WWW-Authenticate response.
+   * This transforms a string value like:
+   * `Bearer authorization="some_authorization", resource="https://some.url"`
+   * into an object like:
+   * `{ authorization: "some_authorization", resource: "https://some.url" }`
+   * @param wwwAuthenticate string value in the WWW-Authenticate header
+   */
+  public parseWWWAuthenticate(wwwAuthenticate: string): ParsedWWWAuthenticate {
+    // First we split the string by either `, ` or ` `.
+    const parts = wwwAuthenticate.split(/,* +/);
+    // Then we only keep the strings with an equal sign after a word and before a quote.
+    // also splitting these sections by their equal sign
+    const keyValues = parts.reduce(
+      (parts, str) => (str.match(/\w="/) ? [...parts, str.split("=")] : parts),
+      [] as string[][]
+    );
+    // Then we transform these key-value pairs back into an object.
+    const parsed = keyValues.reduce(
+      (result, [key, value]: string[]) => ({
+        ...result,
+        [key]: value.slice(1, -1)
+      }),
+      {} as ParsedWWWAuthenticate
+    );
+    return parsed;
+  }
+
+  /**
+   * Parses the given WWW-Authenticate header, generates a new AuthenticationChallenge,
+   * then if the challenge is different from the one cached, resets the token and forces
+   * a re-authentication, otherwise continues with the existing challenge and token.
+   * @param wwwAuthenticate Value of the incoming WWW-Authenticate header.
+   * @param webResource Ongoing HTTP request.
    */
-  public async sendRequest(
+  private async regenerateChallenge(
+    wwwAuthenticate: string,
     webResource: WebResource
   ): Promise<HttpOperationResponse> {
-    if (!webResource.headers) webResource.headers = new HttpHeaders();
+    // The challenge based authentication will contain both:
+    // - An authorization URI with a token,
+    // - The resource to which that token is valid against (also called the scope).
+    const parsedWWWAuth = this.parseWWWAuthenticate(wwwAuthenticate);
+    const authorization = parsedWWWAuth.authorization!;
```

Comment:
feels weird to do all these ! operations to assert something exists. If we know it exists, why make it optional on the type? If we don't know that it exists, why not add some checks so we don't blow up at runtime?

Reply:
I'll add an exception case if we don't receive the values that we expect! Thank you!

Reply:
I added this:

```ts
    if (!(authorization && resource)) {
      return this._nextPolicy.sendRequest(webResource);
    }
```